compiler/hash-ast/src/ast.rs

```diff
@@ -1411,7 +1414,7 @@ define_tree! {
     #[node]
     pub struct StructDef {
         /// Type parameters that are attached to the definition.
-        pub ty_params: Children!(Param),
+        pub ty_params: OptionalChild!(TyParams),
```

Comment:
Shouldn't all of these be just Params??

Reply:
Well, that would mean it uses `<name: ty? = <expr>?` which for type parameters isn't quite right. Also, for something like:
```rs
Foo := <T> => struct(...)
```
would mean the same as:
```rs
Foo := struct<T>(...)
```
So would it not make sense to use a type parameter here?

Reply:
You are right I am just stupid.. But I think TupleTy should use Params like FnTy

Reply:
Also for the second thing you said, should
the first syntax be supported?

Reply:
Whilst this concern is directly related, the question of to which items apply attributes and macros. 

If we decide to simplify the attribute application rules (as in they just apply on whatever item you specified them on), then it would be nice to avoid having something like `<T> => struct(..)` because it kind of complicates what happens in say this case:

```rs
Baz := #foo <T> => struct(..)
```
In this case, the user clearly means to apply `foo` on the struct definition, not the type function itself.

Reply:
True, though this is already an issue with functions anyway. There is a potential solution to this. What if each parameter also allows type arguments? E.g. we have some kind of single Params node, which appears in structs, enums, mods, functions.

And the syntax is

```
params ::= <ident: ty (= ty)?...> (ident: ty (= expr)?...)
```

Both parts () and <> are optional (recovering the original type functions and normal functions). And then the syntax for other items becomes:

```
struct ::= "struct" params fields
enum ::= "enum" params variants
function ::= params "->" ty "=>" expr
function_ty ::= 
...
```

Then there is only one kind of function and you can do something like

```
empty_list := <T> () => List<T>(data=[])

identity := <T> (t: T) -> T => t
```

What do you think about this? It would solve the attribute issue because now generics are part of the same node as the function, not an entirely separate function.

Reply:
It would require some changes to the parser though, idk if you think it is worth it. and I don't know if there are any ambiguities with it.

Reply:
I think we should keep it as it is, and maybe later we can re-visit the subject.

Reply:
> For things like `enum` and `struct` and others, it doesn't make sense for them to have the normal parameter bit though, you'd end up having to write:
> ```rs
> Foo := <T>()(..)
> ```

Yeah in general for simple types it wouldn't be too useful (of course you can omit the empty () the base syntax would be unchanged. You can still write `Foo := struct <T> (field: i32)`).

But this is useful for, e.g. the equality type:

```
Equal := enum <T> (a: T, b: T) (
    Refl(x: T): Equal<T>(a, b)
)

/// Here the T is inferred
transitivity : <T, a: T, b: T, c: T> -> (Equal(a, b), Equal(b, c)) -> Equal(a, c)
                  = ...
```

In general for dependent types with indices it is useful because otherwise you have to pack all the arguments in the <> and then inference is not that nice.

But at the same time this specific thing is not strictly necessary, basically is just an added benefit. The main point is that attributes on functions would not be problematic anymore.

Reply:
Yeah I mean, I'm not really suggesting to do this now, I'm just trying to find a solution to the attribute issue (plus it simplifies the visual noise a bit)